### PR TITLE
systemd-sleep: use resume_offset value if set

### PR DESCRIPTION
**Pull request by zachsmith**
When using a `swapfile` for hibernate, use the `resume_offset` value that has been set. If no `resume_offset` value is found, attempt to calculate as before.

Fixes https://github.com/systemd/systemd/issues/11939

### CODE REVIEW

File: src/sleep/sleep.c [hunk 1]
```diff
@@ -62,15 +62,31 @@ static int write_hibernate_location_info(void) {
...
+        r = read_one_line_file("/sys/power/resume_offset", &resume_offset);
+        if (r < 0)
+                log_warning_errno(r, "Error reading from /sys/power/resume_offset: %m");
+
+        r = safe_atou64(resume_offset, &offset);
+        if (r < 0)
+                log_warning_errno(r, "Unable to verify resume_offset; ignoring");
+        else if (offset > 0) {
+                log_debug("Using /sys/power/resume_offset value: %s", resume_offset);
+                return 0;
```
Comment (keszybz) [inline src/sleep/sleep.c]: When reading something and converting it to a number, we always use the converted value in messages. This has the advantage that in case our conversion is somehow confused, we will report the value that will actually be used later. In other words, something like `log_debug("...: %d", offset)` would be better.

File: src/sleep/sleep.c [hunk 2]
```diff
@@ -62,15 +62,31 @@ static int write_hibernate_location_info(void) {
...
-                return log_debug_errno(errno, "/sys/power/resume_offset not writeable: %m");
+                return log_debug_errno(errno, "/sys/power/resume_offset not readable: %m");
         }
 
+        /* use resume_offset set by user if available */
+        r = read_one_line_file("/sys/power/resume_offset", &resume_offset);
+        if (r < 0)
+                log_warning_errno(r, "Error reading from /sys/power/resume_offset: %m");
+
+        r = safe_atou64(resume_offset, &offset);
```
Comment (keszybz) [inline src/sleep/sleep.c]: I don't understand what this patch is supposed to be doing. `offset` is read here, but it is overwritten below in line 104. The comment in TODO that is removed talks about the kernel command line, while `/sys/power/resume_offset` is read here. I'm confused.

Comment (zachsmith) [reply src/sleep/sleep.c]: If `offset` is `> 0` we return on line 84 thus `offset` is only overwritten if we need to calculate an offset value. Would it be preferable to use a separate variable? That was the way I initially implemented it but seemed like `offset` could be safely reused so that was what I pushed.

I wrote a bit about the reason for for reading from `/sys/power/resume_offset` instead of the kernel cmdline in the bug [here](https://github.com/systemd/systemd/issues/11939#issuecomment-500185206) but in short, I don't think inspecting the cmdline values is the right solution here and that reading the most recent values from `/sys/power/resume_offset` is a better approach since A) the value set on the kernel cmdline will be already be present in `/sys/power/resume_offset` unless B) the value was changed after boot in which case the most recent value will be present in `/sys` and the cmdline would value could be invalid.

Similarly, I don't _think_ there is any need to check the `resume=` kernel cmdline or check `/sys/power/resume` explicitly because if a non-zero value is detected in `/sys/power/resume_offset`, we return before attempting to determine the `device` and write it to `/sys/power/resume` and instead rely on the current swap configuration present at the time of the hibernation request.

This approach implicitly honours whatever the user has configured and if the user has configured incorrectly, the hibernation will fail and log appropriately.

Happy to reconsider this if I'm thinking about it wrong or if there would be better ways to implement this.

Comment (keszybz) [reply src/sleep/sleep.c]: You're right, I missed the `return 0`. So that is fine.

> I wrote a bit about the reason for for reading from /sys/power/resume_offset instead of the kernel cmdline in the bug here but in short ...

Please add the reasoning for commits to the commit description. A straightforward copy&paste is better than nothing. Commits are permanent and easy to find, and we want to be able to check know the reasoning potentially many years in the future. Finding the discussion in a related bug might not be easy at all. (Also, the comment here does say "why", but https://github.com/systemd/systemd/issues/11939#issuecomment-500185206 just describes "what" the code does. So this new comment is much more useful ;)).

> I don't think inspecting the cmdline values is the right solution here and that reading the most recent values from /sys/power/resume_offset is a better approach since A) the value set on the kernel cmdline will be already be present in /sys/power/resume_offset unless B) the value was changed after boot in which case the most recent value will be present in /sys and the cmdline would value could be invalid.

Agreed.

> Similarly, I don't think there is any need to check the resume= kernel cmdline or check /sys/power/resume explicitly because if a non-zero value is detected in /sys/power/resume_offset, we return before attempting to determine the device and write it to /sys/power/resume and instead rely on the current swap configuration present at the time of the hibernation request.

Agreed.

> This approach implicitly honours whatever the user has configured and if the user has configured incorrectly, the hibernation will fail and log appropriately.

Not entirely. So the problem with this patch is that now the flow becomes very complicated:
1. first we call find_hibernate_location()
2. if we found a partition then write to /sys/power/resume and return
3. then we do the resume_offset check you added an potentially return
4. we find an offset using read_fiemap()
5. we write /sys/power/resume_offset and /sys/power/resume.

Now "partition" and "file" are handled very differently, and I don't think this is good. We also check for stuff, potentially failing, and then discard the results. I think the general approach of allowing /sys/power/resume and /sys/power/resume_offset to override our logic is appropriate. But the flow should be:
1. patch find_hibernate_location() to look at /sys/power/resume. If populated, just use that, otherwise look at /proc/swaps as now.
2. if the resume device is a file, check the resume_offset: if specified in /sys/power/resume_offset, use that, otherwise calculate using read_fiemap().
3. write /sys/power/resume_offset (but only if 0 before) and /sys/power/resume (only if empty before).

This fixes some issues:
- we get the same flow for "file" and "partition" (apart from the handling of the resume_offset)
- the check in enough_swap_for_hibernation() now actually checks the device that we'll later use to hibernate ;)

As a bonus point (this probably should be separate commit), we should refuse to look for the offset using read_fiemap() if the file is on btrfs. We know the result will be wrong.

File: src/sleep/sleep.c [hunk 3]
```diff
@@ -62,15 +62,31 @@ static int write_hibernate_location_info(void) {
...
+        if (r < 0)
+                log_warning_errno(r, "Unable to verify resume_offset; ignoring");
```
Comment (keszybz) [inline src/sleep/sleep.c]: Maybe say "The value in /sys/power/resume_offset is not a number, ignoring: %m". (Current message is confusing because it's not clear whether this is about the kernel command line parameter or the sysfs attribute.)

File: src/sleep/sleep.c (hunk 2 above)
Comment (poettering) [inline src/sleep/sleep.c]: please, let's not try to parse `resume_offset` if it's not initialized anyway since read_one_line_file() failed, while this will just fail, it's still ugly, since this means we create two error messages here, and it really looks like a real bug that way

File: src/sleep/sleep.c [hunk 4]
```diff
@@ -62,15 +62,31 @@ static int write_hibernate_location_info(void) {
                                        "Invalid hibernate type: %s", type);
 
         /* Only available in 4.17+ */
-        if (access("/sys/power/resume_offset", W_OK) < 0) {
+        if (access("/sys/power/resume_offset", R_OK) < 0) {
```
Comment (poettering) [inline src/sleep/sleep.c]: there's no need to do the access() check if we open the file right-away here anyway. i mean, if the file doesn#t exist, then read_one_line_file() will return ENOENT, there's no point in checking that beforehand.

The reason the access() check was in place here is to avoid doing all the fiemap work on the raw device (which is security problematic and slow, and ...) if we don't have to.

File: TODO [hunk 5]
```diff
@@ -33,10 +33,6 @@ Features:
   with a nice speed-up on services that have many processes running in the same
   cgroup.
 
-* clean up sleep.c:
-  - Make sure resume= and resume_offset= on the kernel cmdline always take
-    precedence
```
Comment (poettering) [inline TODO:61]: hmm, what about the "resume=" part of this TODO list item, did that already get fixed?

Comment (keszybz) [reply TODO:61]: See my comment above: https://github.com/systemd/systemd/pull/12760#pullrequestreview-250605571.

File: src/sleep/sleep.c (hunk 2 above)
Comment (zachsmith) [reply src/sleep/sleep.c]: Thanks for the great feedback, @keszybz - this approach makes sense to me. I agree this will simplify the flow, put the checks in the right place, and make the logic easier to follow. I'll work on reworking this patch soon.

File: TODO (hunk 5 above)
Comment (zachsmith) [reply TODO:61]: @keszybz and @poettering I refactored my patch based on your feedback. This patch has a much smaller footprint on existing code (just a few debug statements added to the existing calculation code) and adds one new function new function called `configure_hibernation` in `sleep.c` which isolates the configuration precedence logic (`/sys/power` before `/proc/swaps`). Look forward to your feedback!

@keszybz I added a more robust commit statement - please let me know if you were hoping to see something different.

File: src/sleep/sleep.c [hunk 6]
```diff
@@ -141,6 +146,38 @@ static int write_state(FILE **f, char **states) {
...
+        /* check for proper hibernation configuration */
+        r = read_one_line_file("/sys/power/resume", &resume);
+        if (r < 0)
+                return log_debug_errno(r, "Error reading from /sys/power/resume: %m");
+
+        r = read_one_line_file("/sys/power/resume_offset", &resume_offset);
+        if (r < 0)
+                return log_debug_errno(r, "Error reading from /sys/power/resume_offset: %m");
+
+        if ((resume_offset && !streq(resume_offset, "0")) && (resume && !streq(resume, "0:0"))) {
```
Comment (poettering) [inline src/sleep/sleep.c]: given that on failure to read the two files you exit early from this function it's guaranteed that the two variables are non-NULL when you reach this, hence the explicit checks for non-NULL-ness can be dropped.

File: src/sleep/sleep.c [hunk 7]
```diff
@@ -141,6 +146,38 @@ static int write_state(FILE **f, char **states) {
...
+                log_debug("Hibernating using device id and offset read from /sys/power/resume: %s and /sys/power/resume_offset: %s", resume, resume_offset);
+                return 0;
+        }
+        else if (resume && !streq(resume, "0:0")) {
```
Comment (poettering) [inline src/sleep/sleep.c]: we merge the closing } and the `else` line into one usually, please do so here too:

```c
} else if …
```

File: src/sleep/sleep.c [hunk 8]
```diff
@@ -141,6 +146,38 @@ static int write_state(FILE **f, char **states) {
...
+                log_debug("Hibernating using device id read from /sys/power/resume: %s", resume);
+                return 0;
+        }
+        else if (resume_offset && !streq(resume_offset, "0"))
```
Comment (poettering) [inline src/sleep/sleep.c]: ditto, and ditto

File: src/sleep/sleep.c [hunk 9]
```diff
@@ -141,6 +146,38 @@ static int write_state(FILE **f, char **states) {
...
+                log_debug("Found offset in /sys/power/resume_offset: %s; no device id found in /sys/power/resume; ignoring offset", resume_offset);
+
+        /* if hibernation is not properly configured, attempt to calculate and write values */
+        r = write_hibernate_location_info();
+        if (r < 0)
+                return log_error_errno(r, "Failed to write hibernation location information: %m");
```
Comment (poettering) [inline src/sleep/sleep.c]: if write_hibernate_location_info() fails, then we'll log twice here above LOG_DEBUG, as this function will and the calling function does again.

Please make sure that for each error we log exactly once above LOG_DEBUG level. (logging on debug level is generally always fine)

File: src/sleep/sleep.c [hunk 10]
```diff
@@ -166,9 +203,10 @@ static int execute(char **modes, char **states) {
 
         /* Configure the hibernation mode */
         if (!strv_isempty(modes)) {
-                r = write_hibernate_location_info();
+                r = configure_hibernation();
                 if (r < 0)
-                        return log_error_errno(r, "Failed to write hibernation disk offset: %m");
+                        return log_error_errno(errno, "Failed to prepare for hibernation: %m");
```
Comment (poettering) [inline src/sleep/sleep.c]: our functions are supposed to return negative errno-values as return value, and not via the actual `errno` variable. This change which replaces `r` by `errno` is hence not right.

See https://systemd.io/CODING_STYLE#error-handling for details.

File: src/sleep/sleep.c [hunk 11]
```diff
@@ -78,6 +79,14 @@ static int write_hibernate_location_info(void) {
...
                 return log_debug_errno(errno, "Unable to stat %s: %m", device);
 
+        r = btrfs_is_filesystem(fd);
+        if (r < 0)
+                return log_error_errno(r, "Error checking %s for Btrfs filesystem", device);
+
+        if (r)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO),
+                                       "Unable to calculate swapfile offset when using Btrfs: %s", device);
+
```
Comment (zachsmith) [inline src/sleep/sleep.c:91]: > As a bonus point (this probably should be separate commit), we should refuse to look for the offset using read_fiemap() if the file is on btrfs. We know the result will be wrong.

@keszybz here is my attempt at the bonus point as a separate commit :wink:

File: src/sleep/sleep.c [hunk 12]
```diff
@@ -141,6 +146,32 @@ static int write_state(FILE **f, char **states) {
...
+                return log_debug_errno(r, "Error reading from /sys/power/resume: %m");
+
+        r = read_one_line_file("/sys/power/resume_offset", &resume_offset);
+        if (r < 0)
+                return log_debug_errno(r, "Error reading from /sys/power/resume_offset: %m");
+
+        if (!streq(resume_offset, "0") && !streq(resume, "0:0")) {
+                log_debug("Hibernating using device id and offset read from /sys/power/resume: %s and /sys/power/resume_offset: %s", resume, resume_offset);
+                return 0;
+        } else if (resume && !streq(resume, "0:0")) {
```
Comment (poettering) [inline src/sleep/sleep.c]: as in the earlier review: we know for sure that resume is non-NULL, since the two read_one_line_file() calls above succeeded. No need to chec this here again

File: src/sleep/sleep.c [hunk 13]
```diff
@@ -141,6 +146,32 @@ static int write_state(FILE **f, char **states) {
...
+                log_debug("Hibernating using device id read from /sys/power/resume: %s", resume);
+                return 0;
+        } else if (resume_offset && !streq(resume_offset, "0"))
```
Comment (poettering) [inline src/sleep/sleep.c]: ditto

File: src/sleep/sleep.c [hunk 14]
```diff
@@ -78,7 +79,14 @@ static int write_hibernate_location_info(void) {
         if (r < 0)
                 return log_debug_errno(errno, "Unable to stat %s: %m", device);
 
-        // TODO check for btrfs and fail if offset is not provided; calculation will fail
+        r = btrfs_is_filesystem(fd);
+        if (r < 0)
+                return log_error_errno(r, "Error checking %s for Btrfs filesystem", device);
```
Comment (poettering) [inline src/sleep/sleep.c]: please include the error string in the message, i.e. suffix the format string with `: %m`

File: src/sleep/sleep.c [hunk 15]
```diff
@@ -78,7 +79,14 @@ static int write_hibernate_location_info(void) {
...
+
+        if (r)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO),
```
Comment (poettering) [inline src/sleep/sleep.c]: not that it matters much, but I figure EOPNOTSUPP sounds more appropriate here than EIO

File: src/sleep/sleep.c (hunk 12 above)
Comment (zachsmith) [reply src/sleep/sleep.c]: Sorry, I thought I had resolved all of them :disappointed:  - fixed how for sure.

File: src/sleep/sleep.c (hunk 15 above)
Comment (zachsmith) [reply src/sleep/sleep.c]: Thanks for this suggestion. I reviewed `errno-base.h` and selected `EIO` which seemed like the best in that list but I see now that `errno.h` has a much more complete list. If you have any advice or a source I could review for how to select appropriate `errno`'s, I'd appreciate that.

Comment (keszybz) [reply src/sleep/sleep.c]: You can use `errno -l` (the command).

Comment (zachsmith) [reply src/sleep/sleep.c]: And apparently `man errno`... d'oh! :disappointed_relieved:   Thanks!

**Comment by evverx** [comment]
Regarding bionic-i386, I temporarily turned it on in https://github.com/systemd/systemd/pull/12892 just to make sure https://github.com/systemd/systemd/issues/12891 is gone. Looks like this PR was lucky enough to slip through the window as well :-) It can safely be ignored.

**Comment by zachsmith** [comment]
> Regarding bionic-i386, I temporarily turned it on in #12892 just to make sure #12891 is gone. Looks like this PR was lucky enough to slip through the window as well :-) It can safely be ignored.

Should the tests be re-run for these PRs after the bioinc-i386 is turned off?

**Comment by evverx** [comment]
@zachsmith I don't think it's necessary but now that #12760 is merged and bionic-i386 is on again it should be possible to rebase the PR on top of master and force-push the code to retrigger the CI to make the PR "green".

**Comment by zachsmith** [comment]
> @zachsmith I don't think it's necessary but now that #12892 is merged and bionic-i386 is on again it should be possible to rebase the PR on top of master and force-push the code to retrigger the CI to make the PR "green".

Done. Thanks, @evverx!

File: src/sleep/sleep.c [hunk 16]
```diff
@@ -143,6 +156,32 @@ static int write_state(FILE **f, char **states) {
...
+                log_debug("Hibernating using device id and offset read from /sys/power/resume: %s and /sys/power/resume_offset: %s", resume, resume_offset);
+                return 0;
+        } else if (!streq(resume, "0:0")) {
+                log_debug("Hibernating using device id read from /sys/power/resume: %s", resume);
+                return 0;
+        } else if (!streq(resume_offset, "0"))
+                log_debug("Found offset in /sys/power/resume_offset: %s; no device id found in /sys/power/resume; ignoring offset", resume_offset);
+
+        /* if hibernation is not properly configured, attempt to calculate and write values */
+        return write_hibernate_location_info();
```
Comment (keszybz) [inline src/sleep/sleep.c:182]: I like this. The new logic is pretty clear: `resume_offset` is subordinate to `resume` (the device) so we use one or both if `resume` is populated, and calculate the pair ourselves otherwise.

Comment (zachsmith) [reply src/sleep/sleep.c:182]: Thanks for all the help getting this into shape, @keszybz and @poettering. I really appreciate your support.

**Comment by helloworld1** [comment]
This is amazing work! Thanks guys!
